compiler/ir/backend.wasm/src/org/jetbrains/kotlin/backend/wasm/ir2wasm/DeclarationGenerator.kt

```diff
@@ -0,0 +1,329 @@
+        if (wasmFunctionType.resultTypes.isNotEmpty() && declaration.body is IrBlockBody) {
+            exprGen.buildUnreachable()
+        }
+
+        context.defineFunction(declaration.symbol, function)
+
+        if (declaration == backendContext.startFunction)
+            context.setStartFunction(function)
+
+        if (declaration.isExported(backendContext)) {
+            context.addExport(
+                WasmExport.Function(
+                    field = function,
+                    name = declaration.name.identifier
+                )
+            )
+        }
+    }
+
+    override fun visitClass(declaration: IrClass) {
+        if (declaration.isAnnotationClass) return
+        val symbol = declaration.symbol
+
+        if (declaration.isInterface) {
+            context.registerInterface(symbol)
+        } else {
+            val nameStr = declaration.fqNameWhenAvailable.toString()
+            val structType = WasmStructDeclaration(
+                name = nameStr,
+                fields = declaration.allFields(irBuiltIns).map {
+                    WasmStructFieldDeclaration(
+                        name = it.name.toString(),
+                        type = context.transformType(it.type),
+                        isMutable = true
+                    )
+                }
+            )
+
+            context.defineStructType(symbol, structType)
+
+            var depth = 2
+            val metadata = context.getClassMetadata(declaration.symbol)
+            var subMetadata = metadata
+            while (true) {
+                subMetadata = subMetadata.superClass ?: break
+                depth++
+            }
+
+            val initBody = mutableListOf<WasmInstr>()
+            val wasmExpressionGenerator = WasmIrExpressionBuilder(initBody)
+
+            val superClass = metadata.superClass
+            if (superClass != null) {
+                val superRTT = context.referenceClassRTT(superClass.klass.symbol)
+                wasmExpressionGenerator.buildGetGlobal(superRTT)
+            } else {
+                wasmExpressionGenerator.buildRttCanon(WasmRefType(WasmHeapType.Simple.Eq))
+            }
+
+            wasmExpressionGenerator.buildRttSub(
+                WasmRefType(WasmHeapType.Type(WasmSymbol(structType)))
+            )
+
+            val rtt = WasmGlobal(
+                name = "rtt_of_$nameStr",
+                isMutable = false,
+                type = WasmRtt(depth, WasmHeapType.Type(WasmSymbol(structType))),
+                init = initBody
+            )
+
+            context.defineRTT(symbol, rtt)
+            context.registerClass(symbol)
+            context.generateTypeInfo(symbol, binaryDataStruct(context.getClassMetadata(symbol)))
+        }
+
+        for (member in declaration.declarations) {
+            member.acceptVoid(this)
+        }
+    }
+
+    private fun binaryDataStruct(classMetadata: ClassMetadata): ConstantDataStruct {
+        val invalidIndex = -1
```

Comment:
What does `invalidIndex` mean? Is it safe?

Reply:
It means beginning of super class list. It is safe and used in runtime.